src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDatabaseMetaData.java

```diff
@@ -602,77 +605,174 @@ private static String EscapeIDName(String inID) throws SQLServerException {
+                    columns.put(3, "TABLE_NAME");
+                    columns.put(4, "COLUMN_NAME");
+                    columns.put(5, "DATA_TYPE");
+                    columns.put(6, "TYPE_NAME");
+                    columns.put(7, "COLUMN_SIZE");
+                    columns.put(8, "BUFFER_LENGTH");
+                    columns.put(9, "DECIMAL_DIGITS");
+                    columns.put(10, "NUM_PREC_RADIX");
+                    columns.put(11, "NULLABLE");
+                    columns.put(12, "REMARKS");
+                    columns.put(13, "COLUMN_DEF");
+                    columns.put(14, "SQL_DATA_TYPE");
+                    columns.put(15, "SQL_DATETIME_SUB");
+                    columns.put(16, "CHAR_OCTET_LENGTH");
+                    columns.put(17, "ORDINAL_POSITION");
+                    columns.put(18, "IS_NULLABLE");
+                    /*
+                     * Use negative value keys to indicate that this column doesn't exist in SQL Server and should just
+                     * be queried as 'NULL'
+                     */
+                    columns.put(-1, "SCOPE_CATALOG");
+                    columns.put(-2, "SCOPE_SCHEMA");
+                    columns.put(-3, "SCOPE_TABLE");
+                    columns.put(29, "SOURCE_DATA_TYPE");
+                    columns.put(22, "IS_AUTOINCREMENT");
+                    columns.put(21, "IS_GENERATEDCOLUMN");
+                    columns.put(19, "SS_IS_SPARSE");
+                    columns.put(20, "SS_IS_COLUMN_SET");
+                    columns.put(23, "SS_UDT_CATALOG_NAME");
+                    columns.put(24, "SS_UDT_SCHEMA_NAME");
+                    columns.put(25, "SS_UDT_ASSEMBLY_TYPE_NAME");
+                    columns.put(26, "SS_XML_SCHEMACOLLECTION_CATALOG_NAME");
+                    columns.put(27, "SS_XML_SCHEMACOLLECTION_SCHEMA_NAME");
+                    columns.put(28, "SS_XML_SCHEMACOLLECTION_NAME");
+
+                    result_pstmt = (SQLServerPreparedStatement) this.connection
+                            .prepareStatement(generateAzureDWSelect(rs, columns));
+                    userRs = (SQLServerResultSet) result_pstmt.executeQuery();
+                    result_pstmt.closeOnCompletion();
+                    userRs.getColumn(5).setFilter(new DataTypeFilter());
+                    userRs.getColumn(7).setFilter(new ZeroFixupFilter());
+                    userRs.getColumn(8).setFilter(new ZeroFixupFilter());
+                    userRs.getColumn(16).setFilter(new ZeroFixupFilter());
+                    userRs.getColumn(23).setFilter(new IntColumnIdentityFilter());
+                    userRs.getColumn(24).setFilter(new IntColumnIdentityFilter());
+                } catch (SQLException e) {
+                    errorOnClose = e;
+                    if (null != result_pstmt) {
+                        result_pstmt.close();
+                    }
+                } finally {
+                    if (null != errorOnClose) {
```

Comment:
Why is it needed in finally block?

Reply:
Because if the error causes the pstmt to close, then calling pstmt.close() in the catch block will throw something like "statement already closed". That causes us to lose the original error that caused the exception.

Reply:
if SQLException is captured in catch block, try block resource (Resultset rs) is automatically closed and since you set result_pstmt to close on completion, it will be closed too, so you can throw exception from the catch block safely.

Reply:
Because the finally block always executes and I don't want to throw null.

Reply:
Ideally, you should be closing `result_pstmt` in try catch block and catch that exception safely, may just not do anything if it occurs. Coz right now if `result_pstmt.close()` throws exception, you are ignoring thrown exception from this statement and throwing another exception which doesn't seem right.

Alternatively you could do `if (null != result_pstmt && !result_pstmt.isClosed())`

Reply:
This change is to circumvent the issue you described. The exception the statement throws on statement.close() isn't descriptive or relevant to the actual reason the try block failed. We can't add result_pstmt to the try-catch because closing the pstmt that generated the resultset the user needs to read from closes the resultset as well.

Reply:
What I meant to say was:
```java             
} catch (SQLException e) {
    if (null != result_pstmt) {
         try{
               result_pstmt.close();
          } catch (SQLException e) { 
               //do nothing but catch it
          }
    }
    throw e;
}
```
OR
```java             
} catch (SQLException e) {
    if (null != result_pstmt && !result_pstmt.isClosed()) {
          result_pstmt.close();
    }
    throw e;
}
```